platforms/nuttx/src/px4/stm/stm32f4/px4io_serial/px4io_serial.cpp

```diff
@@ -294,49 +291,54 @@ ArchPX4IOSerial::_bus_exchange(IOPacket *_packet)
 
 	/* wait for the transaction to complete - 64 bytes @ 1.5Mbps ~426µs */
 	int ret;
+	irqstate_t irqs = enter_critical_section();
 
 	for (;;) {
 		ret = sem_timedwait(&_completion_semaphore, &abstime);
 
 		if (ret == OK) {
 			/* check for DMA errors */
 			if (_rx_dma_status & DMA_STATUS_TEIF) {
-				// stream transfer error, ensure all DMA is also stopped before exiting early
+				/* stream transfer error, ensure all DMA is also stopped before exiting */
 				_abort_dma();
 				perf_count(_pc_dmaerrs);
 				ret = -EIO;
 				break;
-			}
 
-			/* check packet CRC - corrupt packet errors mean IO receive CRC error */
-			uint8_t crc = _current_packet->crc;
-			_current_packet->crc = 0;
+			} else {
+				/* succesful transaction (crc can still fail) */
+				break;
+			}
 
-			if ((crc != crc_packet(_current_packet)) || (PKT_CODE(*_current_packet) == PKT_CODE_CORRUPT)) {
+		} else {
+			if (errno == ETIMEDOUT) {
+				/* something has broken - clear out any partial DMA state and reconfigure */
 				_abort_dma();
-				perf_count(_pc_crcerrs);
-				ret = -EIO;
+				perf_count(_pc_timeouts);
+				perf_cancel(_pc_txns);		/* don't count this as a transaction */
```

Comment:
Translating this The -EIO may be a better result. It says the is a issue with the IO CPU.
Granted you may want better error discrimination for debugging but the above DMA_STATUS_TEIF error is more of an invalid use of the DMA error.

Reply:
I think the diff here is confusing, I did not change the error codes.

- -1 for timeout
- -EIO for dma error and crc error

Is that what you mean?

Reply:
Yep Confused by the diff.